rasa/core/lock_store.py

```diff
@@ -329,3 +352,114 @@ def _load_from_module_name_in_endpoint_config(
+        if key_prefix:
+            logger.debug(f"Setting non-default redis key prefix: '{key_prefix}'.")
+            self._set_key_prefix(key_prefix)
+
+        super().__init__()
+
+    def _set_key_prefix(self, key_prefix: Text) -> None:
+        if isinstance(key_prefix, str) and key_prefix.isalnum():
+            self.key_prefix = (
+                key_prefix + ":" + DEFAULT_CONCURRENT_REDIS_LOCK_STORE_KEY_PREFIX
+            )
+        else:
+            logger.warning(
+                f"Omitting provided non-alphanumeric redis key prefix: '{key_prefix}'. "
+                f"Using default '{self.key_prefix}' instead."
+            )
+
+    def get_lock(self, conversation_id: Text) -> Optional[TicketLock]:
+        """Retrieves lock (see parent docstring for more information)."""
+        serialised_lock = self.red.get(self.key_prefix + conversation_id)
+        if serialised_lock:
+            return TicketLock.from_dict(json.loads(serialised_lock))
+
+        return None
+
+    def delete_lock(self, conversation_id: Text) -> None:
+        """Deletes lock for conversation ID."""
+        deletion_successful = self.red.delete(self.key_prefix + conversation_id)
+        self._log_deletion(conversation_id, deletion_successful)
+
+    def save_lock(self, lock: TicketLock) -> None:
+        """Commit individual tickets, last issued ticket number and lock to storage."""
+        for ticket in lock.tickets:
+            serialised_ticket = ticket.dumps()
+            key = self.key_prefix + lock.conversation_id + ":" + str(ticket.number)
+            self.red.set(name=key, value=serialised_ticket)
+
+        last_issued_ticket_number = lock.last_issued
+        last_issued_key = (
+            self.key_prefix + lock.conversation_id + ":" + "last_issued_ticket_number"
+        )
+        self.red.set(name=last_issued_key, value=last_issued_ticket_number)
+
+        self.red.set(self.key_prefix + lock.conversation_id, lock.dumps())
+
+    def increment_ticket_number(self, lock: TicketLock) -> int:
+        """Uses Redis atomic transaction to increment ticket number."""
+        last_issued_key = (
+            self.key_prefix + lock.conversation_id + ":" + "last_issued_ticket_number"
+        )
+        last_issued_ticket_number = self.red.get(last_issued_key)
+
+        lock_last_issued = lock.last_issued
+
+        if last_issued_ticket_number is None:
+            return lock_last_issued + 1
+
+        if lock_last_issued != last_issued_ticket_number:
+            serialised_ticket = self.red.get(
+                self.key_prefix + lock.conversation_id + ":" + str(lock_last_issued)
+            )
+
+            if serialised_ticket is None:
+                return lock_last_issued + 1
+
+            ticket = Ticket.from_dict(data=json.loads(serialised_ticket))
+            self.red.set(name=last_issued_key, value=ticket.number)
+
+        return self.red.incr(name=last_issued_key)
```

Comment:
> where is the atomic transaction?

It's in the last line 465; the reason why I had to introduce a non-redis incremental operation in lines 452 and 462 was because of edge cases when there is nothing stored yet when receiving an initial first message. Not having these would lead to errors because the `increment_ticket_number` has to be called before the ticket gets issued in `LockStore.issue_ticket`. Let me know if the integration test doesn't make sense, I assume that if that's passing then we achieved the concurrent outcome?

Reply:
From what I can see, the integration test only verifies that this works under concurrency in the sense of an in-process event loop, but the problem that we are trying is more complex. Cause we're talking about two separate processes, potentially running on separate machines, that use the same redis lock store, so more like true multi-threading than single-threaded concurrency. So the integration test makes sense, but only partly covers the bug we're solving (also testing the other scenario will be hard to set up, not sure if that would be in scope for this here).

Can you explain in more detail what the block LL454-463 does? That one I still don't quite understand. I'm trying to wrap my head around whether we can still get into trouble when this happens. And I apologize if this doesn't make sense because I'm not understanding this area of the code well enough:
- two servers A and B that are running rasa retrieve the same last_issued_ticket_number from redis; the value is 14
- only for one of them (let's say A) will this will be identical to `lock_last_issued`; for B, it is 13 since server A handled ticket 14
- server B then loads the serialized ticket for number 13, and sets the redis key to 13
- then server B increments and returns the ticket number 14
- Server A is just running more slowly and increments afterwards, getting back ticket number 15

So if this scenario can happen like I describe, we have given out ticket number 14 twice, because we re-set the value to 13 in between there. Does this make sense? Or maybe I'm misunderstanding what is saved in `lock.last_issued`?

Reply:
> Can you explain in more detail what the block LL454-463 does?

If I recall correctly, I had to add this block because an integration test was failing (`test_concurrent_lock_expiration`). In `LockStore.issue_ticket` method, I first remove all expired tickets from Lock, before running this `increment_ticket_number` method. This check essentially makes sure that the retrieved last ticket number from redis matches the last issued ticket in the Lock, rather than a potentially expired ticket that was still persisted in the redis store.

Reply:
> only for one of them (let's say A) will this will be identical to lock_last_issued; for B, it is 13 since server A handled ticket 14

This is a good scenario to test actually, I assumed both servers would get the same Lock instance back, so `lock_last_issued` would never change. Any pointers on how to test this probably is very welcome, should I use the `multiprocessing` library?

Reply:
> This check essentially makes sure that the retrieved last ticket number from redis matches the last issued ticket in the Lock

But this is not the case when we have several servers running, is it? Because the last ticket number retrieved from Redis may have been retrieved by a different server. Or maybe I'm still not understanding something here about the whole mechanism 😅

Reply:
> This is a good scenario to test actually, I assumed both servers would get the same Lock instance back, so `lock_last_issued` would never change. Any pointers on how to test this probably is very welcome, should I use the `multiprocessing` library?

Yes, should work with any library that supports multi-threading. But I feel I'm still not understanding something. I thought the Lock object is something local to the process / in-process memory, is it not? Is it somehow retrieved from Redis?

Reply:
> Is it somehow retrieved from Redis?

Yep, it is retrieved from Redis [here.](https://github.com/RasaHQ/rasa/blob/31125d2cefa177b96dd76aeaa5eb46b486cff39b/rasa/core/lock_store.py#L254)

Reply:
Based on our conversation, I refactored by setting expiry times on both individual  persisted tickets and also `last_issued_ticket_number` key which needs to be reset whenever it expires: https://github.com/RasaHQ/rasa/pull/11256/commits/25366adc136856f61aa91bb4eb816bb865cb848b